### ISSUE TITLE
Flexible search

### ISSUE THREAD

**Issue by thegreatmichael** [esvit/ng-table#306]
I think that #293 may be getting at this, but how might we have a custom search? (like one that is across multiple columns?) Right now, you have the filter option which can have multiple keys, these are applied conjunctively (AND). Perhaps a simple first attempt at this would allow disjunctive (OR) filtering?

**Comment by faustlife** [comment]
I have the same problem, disjunctive (OR) filtering work for all columns but not for custom....

**Comment by thegreatmichael** [comment]
@faustlife  I don't know if this is best, but i ended up adding the following to the controller where i was using the table:

tl;dr: i wrote a funciton that would take the query and the items and only return the matching results, and i used that in the `getData` function...

    function tableFilter (query, items) {
      var filteredResults = [];
      for (var i=0; i<items.length; i++) {
        var item = items[i];
        for (var prop in item) {
          if (item.hasOwnProperty(prop) && item[prop].toString().toLowerCase().indexOf(query.toLowerCase()) >= 0) {
            filteredResults.push(item);
            break;
          }
        }
      }
      return filteredResults;
    }
    $scope.tableParams = new ngTableParams({
        count: 1000,          // count per page
        sorting: {
            datetime: 'asc'     // initial sorting
        }
      }, {
      counts: [],
      total: $scope.totalOfferings(), 
      getData: function($defer, params) {
        $scope.filteredData = $scope.offerings[$scope.thisTerm.name];
        if ($scope.hasOwnProperty('searchQuery') && $scope.searchQuery.length>0) {
          $scope.filteredData = tableFilter($scope.searchQuery, $scope.filteredData, params.orderBy());    
        }
        var orderedData = params.sorting() ? $filter('orderBy')($scope.filteredData, params.orderBy()) : $scope.filteredData;
        $defer.resolve(orderedData);
      }
    });

**Comment by faustlife** [comment]
@thegreatmichael hah, intresting idea.
I use practicaly the same method, add  new data array with columns for filtering and use standart filter with text value, as a result I have three arrays: all data array {id,name,jobtitle,datereg}, array with filters columns {name,jobtitle} and array with filtering data.

**Comment by faustlife** [comment]
@thegreatmichael if I understand correctly $scope.filteredData = $scope.offerings[$scope.thisTerm.name], $scope.filteredData it's a list objects with fiels for filtering not full objects,but I need in $scope.filteredData lisf of full objects and filtering by objects with fiels for filtering.

**Comment by thegreatmichael** [comment]
For me it's the full objects. I hardcoded into the tablefilter fn above the fields I care about and just do a string comparison. Actually it looks like I check all the fields, so that's why there's no object for defining which fields to search. The `for...in` loop checks all the fields of he actual objects against the query.

**Comment by faustlife** [comment]
@thegreatmichael  I create custom filter it use disjunctive (OR) filtering

app.filter('propsFilter', function() {
  return function(items, props) {
    var out = [];

    if (angular.isArray(items)) {
      items.forEach(function(item) {
        var itemMatches = false;

        var keys = Object.keys(props);
        for (var i = 0; i < keys.length; i++) {
          var prop = keys[i];
          var text = props[prop].toLowerCase();
          if (item[prop].toString().toLowerCase().indexOf(text) !== -1) {
            itemMatches = true;
            break;
          }
        }
        if (itemMatches) {
          out.push(item);
        }
      });
    } else {
      // Let the output be the input untouched
      out = items;
    }
    return out;
  };});

I use it like this:
$scope.filtered_data = params.filter() ? $filter('propsFilter')($scope.data, params.filter().deffilter)
                  : $scope.data; 
it works fine!
